### ISSUE TITLE
How to gracefully stop server programatically

### ISSUE THREAD

**Issue by bitrut** [aio-libs/aiohttp#2950]
## Long story short

I have a aiohtto-based service which has the main purpose to just perform a task running in the background. The http is here just to respond with health status of the background task. I don't see any way to programatically stop aiohttp server in case of exception raised by the task. The whole service should stop, because it doesn't make sense without healthy task. I tried something like this (the example is based on [the aiohttp docs](https://docs.aiohttp.org/en/stable/web_advanced.html#background-tasks)):

```python
import logging
from aiohttp.web_runner import GracefulExit

log = logging.getLogger(__name__)

async def some_task(app):
    try:
        some_long_running_task()
    except asyncio.CancelledError:
        pass
    except Exception:
        log.exception('Some logs')
        raise GracefulExit()
    finally:
        await cleanup()
```

## Expected behaviour

I'd expect aiohttp to run all the `on_cleanup` hooks. Or some documentation on how to programatically shutdown the service.

## Actual behaviour

`on_cleanup` hooks are not called.

**Comment by asvetlov** [comment]
https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.AppRunner.cleanup

**Comment by bitrut** [comment]
FYI, raising `GracefulExit` did the job. I just had some issues with concurrency and I've been raising this exception too early in my testing code.
Thanks anyway.

**Comment by webknjaz** [comment]
@TomGoBravo FYI you may experience intermittent problems when running asyncio loop in non-main threads.

**Comment by conraddd** [comment]
@TomGoBravo 
Same issue here, the application keeps running after i called ```app.shutdown()``` and ```app.cleanup()```

Here is the code

```Python
import asyncio
from aiohttp import web

app = web.Application()
routes = web.RouteTableDef()

@routes.get("/hello")
async def hello(request):
    return web.Response(text="Hello World")


async def bootup(this_app):
    asyncio.create_task(background())

async def background():
    await asyncio.sleep(5)
    print("Start shutting down")
    await app.shutdown()
    print("Start cleaning up")
    await app.cleanup()

if __name__ == "__main__":
    app.add_routes(routes)
    app.on_startup.append(bootup)
    web.run_app(app, host="0.0.0.0", port=80)
    print("Finished")

```
output
```Bash
======== Running on http://0.0.0.0:80 ========
(Press CTRL+C to quit)
Start shutting down
Start cleaning up

```

**Comment by kohtala** [comment]
This seems to be without a solution. I find related issues #5386, #4617, #3638, ...

I think the new API to solve it is in #2746. But the PR is stalled. I did not go studying further.

**Comment by kohtala** [comment]
I solved this by using the `web.AppRunner` and having the main task just wait on a future until the other tasks have decided they have shut down. The runner.cleanup should perform the shutdown. Something along the lines of:

```python
async def process():
    app = web.Application()
    runner = web.AppRunner(app)
    loop = asyncio.get_running_loop()
    running = loop.create_future()

    async def stop(e):
        await runner.cleanup()
        running.set_exception(e)
    ... # app setup
    await runner.setup()
    site = web.TCPSite(runner, *address)
    await site.start()
    await running

def main(config=None):
    asyncio.run(process())
```

Would be nice to have a `web.serve_app` you could just await or cancel with as simple interface as that of `web.run_app` along the lines of `asyncio.Server.serve_forever`.

**Comment by tilsche** [comment]
I kindly ask for this to be reopened.

It is very frustrating that there only seem to be two ways of _causing_ a graceful shutdown of an aiohttp application as discussed in this issue:

1. Raising a `GracefulExit`, having to use a private API
2. Manually re-implementing `web.run_app`, probably by also using private APIs

Even then, it seems questionable just how graceful a shutdown by throwing a `SystemExit` to the loop really is (see https://github.com/aio-libs/aiohttp/issues/3638#issuecomment-621256659). So some documentation would really be appreciated.

**Comment by mgraczyk** [comment]
In my case the easiest way to do this in most cases is to send a signal like `os.kill(os.getpid(), 15)`. You could modify the signal handler if necessary for more specific behavior.

**Comment by nosahama** [comment]
This did the trick for me and also exits gracefully. 
```python
import signal

def shutdown():
    signal.raise_signal(signal.SIGTERM)
```

Call `shutdown()` where you'd like to perform the interception.

**Comment by qris** [comment]
I would also like to request that this be reopened. I run an asyncio server in a thread during a test. When the test finishes, I want to stop it. I really don't know how to do that gracefully. I can get `all_tasks()` and `cancel` each one, but I don't know if that's clean (I suspect not because it throws uncaught exceptions) and I cannot signal my own thread with SIGTERM because it will kill the whole process (and would not work on Windows anyway). Please could we have a nice supported shutdown mechanism? We don't even have access to the `runner` object to `stop` it.

**Comment by Dreamsorcerer** [comment]
Why don't you have access to the runner object? It sounds like the kind of situation I'd expect you to use AppRunner and control the execution there.

**Comment by qris** [comment]
We call:
```
aiohttp.web.run_app(app)
```
Which internally does:
```
    main_task = loop.create_task(_run_app(...))
    ...
    try:
        asyncio.set_event_loop(loop)
        loop.run_until_complete(main_task)
    except (GracefulExit, KeyboardInterrupt):  # pragma: no cover
        pass
    finally:
        ...
```
The only way out of this is to raise `GracefulExit` (a private API, as mentioned above) or `KeyboardInterrupt` (probably by sending a SIGINT to ourselves, or faking it).

The `runner` is created by `_run_app`.

To exit the loop any other way, without using private APIs, I'd need to copy and paste both `run_app` and `_run_app` to make very small changes to them, e.g. add another exception like GracefulExit, or have `_run_app` wait for an asyncio Event rather than doing this:

```
        # sleep forever by 1 hour intervals,
        while True:
            await asyncio.sleep(3600)
```

I really dislike copying large amount of code to make tiny changes to them. If control was inverted here, or even if `GracefulExit` was made pubic, that would not be necessary.

**Comment by Dreamsorcerer** [comment]
As mentioned above, your case sounds like run_app() should not be used (i.e. you are not using it is the entry point of your application, which we are pretty explicit about it being designed for). I don't think you should need to copy code from run_app() at all, if you just use the runners instead, as in:
https://docs.aiohttp.org/en/stable/web_advanced.html#application-runners

Alternatively, a different approach for testing could make sense. Typically we test applications using pytest-aiohttp which handles all this automatically. For testing something like a frontend app which needs the backend running, we'd typically run the backend app in a separate process and kill the process at the end.

I don't think there's really enough demand to focus development on this edge case, but it might be possible to create a public API if someone wanted to work on it that desperately. But, it'd need decent test coverage. Someone would need to think about it, and test it, working within a handler (ideally with the handler still returning), from a separate task outside of a handler etc.
